docs/AUTH_API.md

```diff
@@ -0,0 +1,73 @@
+# Authentication API
+
+For now this is just a dumping ground of ideas how to take the [advanced JSON login protocol](AUTH_PROTOCOL.md) and turn it into an API to be used by the rest of the system. Eventually I expect that this would morph into a doc page that more conherently explains the components I propose here and explains how they all work together to provide a new `systemd-authd` (name TBD?) component.
+
+## `io.systemd.AuthenticationProvider`
+
+This Varlink interface has a single method: `Authenticate(username | uid)`. It returns multiple values:
+
+1. First it sends over the file descriptor. This can be sent off to the frontend
+2. Later, once the frontend has finished authentication, it returns either a `AuthenticationFailed` error or success. In the success case, it returns a blob of data to be used for things like home directory encryption/decryption and whatnot.
+
+The providers are dropped into `/run/systemd/auth-provider/[name]`. You can decide which provider should be used by looking at the `authService` field of the user's user record.
+
+## `/usr/lib/systemd/systemd-authd-pam-bridge`
+
+For various display managers to port to this new mechanism, they need to be able to maintain compatibility with users that don't have an `AuthenticationProvider` associated and potentially some custom PAM authentication plugin (i.e. imagine someone who previously implemented two-factor by having `pam_unix.so` followed by `pam_google_authenticator.so`). To that end, we need to have backwards-compatibility for running things like this.
+
+This is what `systemd-authd-pam-bridge` is for. If we don't have an `AuthenticationProvider` associated with a given user, then each service (GDM, Plasma DM/SDDM, LightDM, etc) can fall back to PAM.
+
+The binary would take one argument, and that's the PAM stack to run. It runs that PAM stack, and then translates the PAM conversation into the new JSON protocol.
+
+If we're feeling adventurous, we could implement some of GDM's PAM extesnsion. Specifically, I've got the GDM Chooser extension in mind. If we're feeling extra adventurous, we could try implementing the various GDM JSON extensions as well (i.e. the authd protocol and SSSD protocol). Though it might be better for SSSD and authd to just port to the new JSON protocol.
+
+We should also support the extension that is used to tunnel the new advanced JSON protocol through PAM. That way we can transparently pass through any PAM modules that are capable of the advanced JSON protocol.
```

Comment:
Yeah, I think we can avoid them unless strictly necessary, but if this proposal gets to an end, ideally I think that we should just enforce other systems to just use this protocol.